### DISCUSSION TITLE
Environment variable expansion

### DISCUSSION THREAD

**Discussion by StefanoDeVuono** [vercel/vercel#5355] [Ideas]
I'd like to add environment variable expansion so that a `.env` file with
```
VARIABLE_1="variable one"
VARIABLE_TO_EXPAND=$REFERENCE_OTHER_ENV_VAR
```
could use set `VARIABLE_TO_EXPAND` as whatever `$REFERENCE_OTHER_ENV_VAR` evaluates to.

I would add `dotenv-expand` to `now-cli/src/util/dev/server.ts`.

What do people think about this?

**Comment by amille14** [comment]
This would be a very welcome change, especially being able to reference system env vars like `VERCEL_URL` in other variables

**Reply by lucleray** [reply]
We recently released [References to System Environment Variables](https://vercel.com/docs/platform/environment-variables#references-to-system-environment-variables) that allows you to do exactly this!

**Reply by ignat** [reply]
It's not the same thing. It's just a copy of the variable, not a variable expansion.

What is needed is this `VAR2="https://${VERCEL_URL}/login/callback"`

**Reply by lucleray** [reply]
This is an interesting idea but I've never seen this syntax in a `.env` file. Is there any existing libraries that support that?

**Reply by ignat** [reply]
[dotenv-expand](https://github.com/motdotla/dotenv-expand)

**Reply by amille14** [reply]
Yes, I second what @ignat said about allowing `dotenv-expand` syntax. It's what we use and would make life a lot easier if vercel supported this.

**Reply by StefanoDeVuono** [reply]
I think if we add `dotenv-expand` to `now-cli/src/util/dev/server.ts` that should help. I wrote a prototype of this somewhere...

**Comment by ignat** [comment]
Yes, [dotenv-expand](https://github.com/motdotla/dotenv-expand) mentioned above.

**Comment by benderillo** [comment]
@lucleray So, it is 2022, I wonder if anything changed here?
I recently needed to compose a release name out of GIT_SHA and a prefix "rel" and I went for creating an env variable like this:
`rel-$VERCEL_GIT_COMMIT_SHA` which did not work. Not to mention that the `rel-${VERCEL_GIT_COMMIT_SHA}` doesn't work either. 
Is there **ANY** way I can define env variable that is constructed out of other env variables on Vercel?

**Reply by chemalopezp** [reply]
By chance I just bumped into this, quite surprised it is not supported by Vercel. As mentioned above `dotenv-expand` is quite standard in order to compose environment variables.

It would be very useful, for example to generate different preview environments (e.g. have a different database name / url for each git branch, etc.)

**Comment by T99** [comment]
Wanted to bump this again as I had just assumed that this was the way it would work on Vercel after it having worked this way in my local environment... my build failed unexpectedly because of it.

Any planned support here? If not, any solutions/hacks from anyone on how to achieve the same effect?

**Reply by T99** [reply]
Ah, OK, so it looks like this is more-so a bug with the Vercel web interface than anything else. If you add your environment variables from a .env using the vercel CLI, then things should work as expected.

More here --> https://github.com/vercel/vercel/discussions/8579
